src/compiler/plugin.ts

```diff
@@ -457,19 +442,12 @@ export namespace CodeGeneratorResponse {
                 insertion_point?: string;
                 content?: string;
                 generated_code_info?: ReturnType<typeof dependency_1.GeneratedCodeInfo.prototype.toObject>;
-            } = {};
-            if (this.name != null) {
-                data.name = this.name;
-            }
-            if (this.insertion_point != null) {
-                data.insertion_point = this.insertion_point;
-            }
-            if (this.content != null) {
-                data.content = this.content;
-            }
-            if (this.generated_code_info != null) {
-                data.generated_code_info = this.generated_code_info.toObject();
-            }
+            } = {
+                name: pb_1.Message.getField(this, 1) != null ? this.name : undefined,
```

Comment:
I am not sure if this is the way to go. given that old Object.keys() and Object.keys() with this change would yield different results. 

I believe the logic 
```ts
if (message.getField(this, 1) != null)  {
  // set it now
}

or am i missing something here?
```

Reply:
I guess it's up the preference. It might introduce breaking changes tho. For me personally I prefer the new approach.

protoc javascript output for comparsion:
```js
/**
 * Static version of the {@see toObject} method.
 * @param {boolean|undefined} includeInstance Deprecated. Whether to include
 *     the JSPB instance for transitional soy proto support:
 *     http://goto/soy-param-migration
 * @param {!proto.google.protobuf.compiler.CodeGeneratorResponse.File} msg The msg instance to transform.
 * @return {!Object}
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.google.protobuf.compiler.CodeGeneratorResponse.File.toObject = function(includeInstance, msg) {
  var f, obj = {
    name: (f = jspb.Message.getField(msg, 1)) == null ? undefined : f,
    insertionPoint: (f = jspb.Message.getField(msg, 2)) == null ? undefined : f,
    content: (f = jspb.Message.getField(msg, 15)) == null ? undefined : f,
    generatedCodeInfo: (f = msg.getGeneratedCodeInfo()) && proto.google.protobuf.GeneratedCodeInfo.toObject(includeInstance, f)
  };

  if (includeInstance) {
    obj.$jspbMessageInstance = msg;
  }
  return obj;
};
}
```

Reply:
I wonder what would be the benefit of setting the property to undefined if it is not there?

Reply:
I guess it mainly comes from other programming languages. That means there wouldn't be any optional property (`?`) but (`type | undefined`). Most of the time people use optional property, because of lazyness.

Now what are the benefits. I guess a big part comes from a readability/debuggability standpoint. You probably know what type of object you're operating on just by having the members there. And secondly let's say you want to dynamically render a table without specifying the columns (property names). I think you can't do that right now?

Thinking about this. It would be good to have interface declarations instead of anonymous types aswell.

Reply:
but doesn't `property?: type` translates to `property: type | undefined`? 

> And secondly let's say you want to dynamically render a table without specifying the columns (property names). I think you can't do that right now?

My personal preference is that you shouldn't do this even if you can (this doesn't count though).  but even if you wanted to you could easily do `Object.keys(message).filter(k => typeof message[k] != "function")`

my second point is that we'd lose the ability to do `'key' in message.toObject()` which I am a big fan of.

I say we don't break the way toObject works now since a lot of people in lots of places could be already depending on this behavior.  We could revisit this idea later and hide behind a feature gate?

Reply:
> but doesn't `property?: type` translates to `property: type | undefined`?
> 

It does but compiler won't let you initialize the object without the property set.

> My personal preference is that you shouldn't do this even if you can (this doesn't count though). but even if you wanted to you could easily do `Object.keys(message).filter(k => typeof message[k] != "function")`
> 

But then you need to keep carrying on the message object. Also you might get some properties (idk if there is any) from the pb1_Message parent itself. And you still won't get unset properties. Also what do you mean you shouldn't? Let's say I fetch list of people (10 properties). Some are set, some are undefined. Then I want to render a table of each person with all 10 columns. And then if dev adds a new property I just want to generate a new model and see the 11th column without modifying of the table code.

I guess the difference is that I look at it from the object perspective while you look at it more like a dictionary? For me the `toObject` should return clear object that doesn't care about protobuf default values nor what is or isn't serialized in the process. Simply optional = give me a nullable type (the `type | undefined` one) otherwise return the value (or default). Can you tell me what benefit do you see in the current approach? Maybe I will get it then.

But either way I don't think it's anything required right now.

Reply:
Thanks for a discussion on the advantages and disadvantages of keeping undefined in fields, @thesayyn and @Yurzel.

My understanding is that I should change the current behaviour of the code in this pull request, so that those undefined fields are NOT present in the objects generated by the `toObject()` method.  The behaviour where those undefined fields are present in the returned objects could be added later as a feature that is implicitly disabled but can be turned on via a command-line option.
@thesayyn, please correct me if I am wrong.

Personally I do not have any personal preference for keeping or omitting the undefined fields, but I agree that keeping them could potentially break some existing code using the library.  I must also admit that it felt somewhat wrong to modify existing tests so that they tolerate the extra undefined fields.  By making the undefined fields disappear from the generated objects we will be able to restore some *.spec.ts files to their previous state (and then the only changes in *.spec.ts will be those related to the default values).

Reply:
> My understanding is that I should change the current behaviour of the code in this pull request, so that those undefined fields are NOT present in the objects generated by the `toObject()` method.

Yes. let's keep the current behavior and revisit the new idea later.

Reply:
> The behaviour where those undefined fields are present in the returned objects could be added later as a feature that is implicitly disabled but can be turned on via a command-line option.
> @thesayyn, please correct me if I am wrong.

Exactly.